Include sources from composite builds in Groovydoc

jamesfredley opened the pull request:
Updated the Groovydoc configuration to collect source directories from included (composite) builds by inspecting the file system. This ensures documentation generation includes sources from all relevant projects, and adds exclusions to prevent duplicate class errors.

Remove `grails.util.BuildScope` documentation
`grails.util.BuildScope` was removed in Grails 4

jamesfredley committed:
Include sources from composite builds in Groovydoc

Updated the Groovydoc configuration to collect source directories from included (composite) builds by inspecting the file system. This ensures documentation generation includes sources from all relevant projects, and adds exclusions to prevent duplicate class errors.

jamesfredley committed:
Remove grails.util.BuildScope documentation

grails.util.BuildScope was removed in Grails 4

jamesfredley committed:
Update API docs link in README

Changed the API documentation link in the README to point to https://docs.grails.org/latest/api/ instead of the previous GitHub Pages URL.

matrei commented on grails-doc/src/en/guide/plugins/understandingPluginLoadOrder.adoc:
```diff
@@ -99,23 +99,14 @@ It's not only plugin load order that you can control. You can also specify which
 ----
 
-In this example, the plugin will only load in the 'development' and 'test' environments. Nor will it be packaged into the WAR file, because it's excluded from the 'war' phase. This allows `development-only` plugins to not be packaged for production use.
+In this example, the plugin will only load in the 'development' and 'test' environments.
```
Will it not also load in the `myCustomEnv` environment?

jamesfredley replied:
You are correct, the historical docs did not have that for some reason.  Adding it.

jdaugherty commented on grails-doc/build.gradle:
```diff
@@ -82,13 +82,36 @@ combinedGroovydoc.configure { Groovydoc gdoc ->
+    List<File> includedBuildSourceDirs = gradle.includedBuilds
+            .collectMany { includedBuild ->
+                def sourceDirs = []
+                includedBuild.projectDir.eachDirRecurse { File dir ->
```
Won't this recurse into every descendant directory while you only need 2 of these?  Seems like we should optimize this so it doesn't slow down the build.

jamesfredley replied:
.eachDir was sufficient.  Updated.

jdaugherty commented on grails-doc/build.gradle:
```diff
@@ -82,13 +82,36 @@ combinedGroovydoc.configure { Groovydoc gdoc ->
+
+    gdoc.source(project.files((baseSourceDirs + includedBuildSourceDirs).unique()))
+    // Exclude files that are not part of the public API and cause duplicate class errors when combined with other source sets
+    gdoc.exclude('org/grails/example/**', 'org.grails.example/**', 'another/**', 'TestJava**')
```
These packages look like test packages, but tests should not be in the groovydoc.  Do you have an example of which file was causing this problem?  This doesn't seem right to me.

jamesfredley replied:
Entry org/grails/example/SubProject1.groovy is a duplicate but no duplicate handling strategy has been set.
Entry org.grails.example/SubProject2.groovy is a duplicate but no duplicate handling strategy has been set.
Entry another/TestOtherJava.java is a duplicate but no duplicate handling strategy has been set.
Entry TestJava.java is a duplicate but no duplicate handling strategy has been set.

jamesfredley replied:
This was resolved by going from eachDirRecurse to .eachDir.  That's where the pollution was coming from.

jamesfredley committed:
Fix environment list in plugin load order example

Corrects the documentation to state that the plugin will load in 'development', 'test', and 'myCustomEnv' environments, matching the provided code example.

jamesfredley committed:
Update Groovydoc source dir traversal to non-recursive

Changed directory traversal from recursive to non-recursive when collecting source directories for Groovydoc.
